modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1952,50 +1952,56 @@ public void onError(Throwable e) {
             @Override
             public void onUpdate(WatchEvent evt) {
-                if (!busyLock.enterBusy()) {
-                    throw new IgniteInternalException(new NodeStoppingException());
-                }
-
-                try {
-                    assert evt.single();
+                inBusyLock(busyLock, () -> {
+                    assert evt.single() : evt;
 
                     Entry stableAssignmentsWatchEvent = evt.entryEvent().newEntry();
 
                     if (stableAssignmentsWatchEvent.value() == null) {
                         return;
                     }
 
-                    int part = extractPartitionNumber(stableAssignmentsWatchEvent.key());
-                    UUID tblId = extractTableId(stableAssignmentsWatchEvent.key(), STABLE_ASSIGNMENTS_PREFIX);
+                    int partitionId = extractPartitionNumber(stableAssignmentsWatchEvent.key());
+                    UUID tableId = extractTableId(stableAssignmentsWatchEvent.key(), STABLE_ASSIGNMENTS_PREFIX);
 
-                    TablePartitionId replicaGrpId = new TablePartitionId(tblId, part);
+                    TablePartitionId replicaGrpId = new TablePartitionId(tableId, partitionId);
 
                     Set<Assignment> stableAssignments = ByteUtils.fromBytes(stableAssignmentsWatchEvent.value());
 
-                    byte[] pendingFromMetastorage = metaStorageMgr.get(pendingPartAssignmentsKey(replicaGrpId),
-                            stableAssignmentsWatchEvent.revision()).join().value();
+                    byte[] pendingAssignmentsFromMetaStorage = metaStorageMgr.get(
+                            pendingPartAssignmentsKey(replicaGrpId),
+                            stableAssignmentsWatchEvent.revision()
+                    ).join().value();
 
-                    Set<Assignment> pendingAssignments = pendingFromMetastorage == null
+                    Set<Assignment> pendingAssignments = pendingAssignmentsFromMetaStorage == null
                             ? Set.of()
-                            : ByteUtils.fromBytes(pendingFromMetastorage);
+                            : ByteUtils.fromBytes(pendingAssignmentsFromMetaStorage);
 
                     String localMemberName = clusterService.topologyService().localMember().name();
 
                     boolean shouldStopLocalServices = Stream.concat(stableAssignments.stream(), pendingAssignments.stream())
                             .noneMatch(assignment -> assignment.consistentId().equals(localMemberName));
 
                     if (shouldStopLocalServices) {
+                        TableImpl table = tablesByIdVv.latest().get(tableId);
+
+                        CompletableFuture<Void> destroyMvPartitionStorageFuture
+                                = table.internalTable().storage().destroyPartition(partitionId);
+
+                        table.internalTable().txStateStorage().destroyTxStateStorage(partitionId);
+
+                        // Necessary evil.
```

Comment:
Are you sure that it's necessary?
Why do we need to await the destruction?
If it's only about awaiting on component stop, it's possible to await allOf storage destruction futures before components stop itself.

Reply:
Methods `TxStateTableStorage#destroyTxStateStorage` and `MvTableStorage#destroyPartition` close the storages and clean up the data.
Thus, if we close the storages, then in order to clear them, we will need to create them again, which may not look good.
But in fact, deleting storages is done in almost O(1) at the moment.
I can describe it in the comments so that there are fewer questions.

Reply:
> Thus, if we close the storages, then in order to clear them, we will need to create them again, which may not look good.

My point is that in case of storage recreation we should chain it with storage.drop(). 
And besides that we should await storage drop on components stop.

Reply:
The problem of clearing the storages when the node is restarted will be solved [IGNITE-18160](https://issues.apache.org/jira/browse/IGNITE-18160) when the node is stopped, we do not need to wait for the data to be cleared, it can take a long time.

Reply:
And what about storage recreation?
What I want to say is that if we want to drop the storage with some id X we can put destroyMvPartitionStorageFuture to the map using X as a key, and if we try to recreate the storage with same id we can chain storage creation with storage drop.

> we do not need to wait for the data to be cleared, it can take a long time.

That's a bit confusing. You've told that it's O(1)

> But in fact, deleting storages is done in almost O(1)

Reply:
Discussed personally, decided to do it through an external map with destruction futures.

We will wait for them at the re-creation of storages.
As well as at stopping nodes.